overlord/ifacestate/helpers.go

```diff
@@ -540,3 +540,167 @@ func resolveSnapIDToName(st *state.State, snapID string) (name string, err error
 	}
 	return decl.SnapName(), nil
 }
+
+// InterfaceMapper offers APIs for re-mapping plugs and slots in interface
+// connections. The mapper is designed to apply transformations around the
+// edges of snapd (state interactions and API interactions) to offer one view
+// on the inside of snapd and another view on the outside.
+type InterfaceMapper interface {
+	RemapIncomingPlugRef(plugRef *interfaces.PlugRef) (changed bool)
+	RemapOutgoingPlugRef(plugRef *interfaces.PlugRef) (changed bool)
+	RemapIncomingSlotRef(slotRef *interfaces.SlotRef) (changed bool)
+	RemapOutgoingSlotRef(slotRef *interfaces.SlotRef) (changed bool)
+}
+
+// NilMapper implements InterfaceMapper and performs no transformations at all.
+type NilMapper struct{}
+
+// RemapIncomingPlugRef doesn't change the plug in any way.
+func (m *NilMapper) RemapIncomingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return false
+}
+
+// RemapOutgoingPlugRef doesn't change the plug in any way.
+func (m *NilMapper) RemapOutgoingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return false
+}
+
+// RemapIncomingSlotRef doesn't change the slot in any way.
+func (m *NilMapper) RemapIncomingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	return false
+}
+
+// RemapOutgoingSlotRef doesn't change the slot in any way.
+func (m *NilMapper) RemapOutgoingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	return false
+}
+
+// mapper contains the currently active interface mapper.
+var mapper InterfaceMapper = &NilMapper{}
+
+// MockInterfaceMapper mocks the currently used interface mapper.
+func MockInterfaceMapper(new InterfaceMapper) (restore func()) {
+	old := mapper
+	mapper = new
+	return func() { mapper = old }
+}
+
+// CoreSnapdMapper implements InterfaceMapper and makes implicit slots
+// appear to be on "core" while they are on "snapd" in reality.
+type CoreSnapdMapper struct{}
+
+// RemapIncomingSlotRef make slots appear on "snapd" rather than "core".
+func (m *CoreSnapdMapper) RemapIncomingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	if slotRef.Snap == "core" {
+		slotRef.Snap = "snapd"
+		return true
+	}
+	return false
+}
+
+// RemapOutgoingSlotRef make slots appear on "core" rather than "snapd".
+func (m *CoreSnapdMapper) RemapOutgoingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	if slotRef.Snap == "snapd" {
+		slotRef.Snap = "core"
+		return true
+	}
+	return false
+}
+
+// RemapIncomingPlugRef doesn't change the plug in any way.
+func (m *CoreSnapdMapper) RemapIncomingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return false
+}
+
+// RemapOutgoingPlugRef doesn't change the plug in any way.
+func (m *CoreSnapdMapper) RemapOutgoingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return false
+}
+
+// Remapping functions for incoming transfers ({state,wire}=>memory)
```

Comment:
it would be good if we found something clearer than outgoing/incoming for the terminology